runtime/src/accounts_db.rs

```diff
@@ -6156,32 +6162,34 @@ impl AccountsDb {
         let len = std::cmp::min(accounts.len(), infos.len());
         let chunk_size = std::cmp::max(1, len / quarter_thread_count()); // # pubkeys/thread
         let batches = 1 + len / chunk_size;
-        (0..batches)
-            .into_par_iter()
-            .map(|batch| {
-                let start = batch * chunk_size;
-                let end = std::cmp::min(start + chunk_size, len);
-                let mut reclaims = Vec::with_capacity((end - start) / 2);
-                (start..end).into_iter().for_each(|i| {
-                    let info = infos[i];
-                    let pubkey_account = (accounts.pubkey(i), accounts.account(i));
-                    let pubkey = pubkey_account.0;
-                    let old_slot = accounts.slot(i);
-                    self.accounts_index.upsert(
-                        target_slot,
-                        old_slot,
-                        pubkey,
-                        pubkey_account.1,
-                        &self.account_indexes,
-                        info,
-                        &mut reclaims,
-                        previous_slot_entry_was_cached,
-                    );
-                });
-                reclaims
-            })
-            .flatten()
-            .collect::<Vec<_>>()
+        self.thread_pool.install(|| {
```

Comment:
This function can be called from both fg (tx processing, bank freeze) and bg processes (clean, shrink). I have a comment above where when I tried to put a thread pool on here we got deadlocks. Perhaps I was trying to use the bg thread pool. I specifically implemented this function to use a controlled number of threads max and avoided a thread pool.

Reply:
We can debug deadlocks if they are reproducible but as it stands, this is still taking a lot of resources; and we have a performance regression on current master code which I am trying to debug.
I see the function is invoked from two paths:
* `store_accounts_frozen` which I can only see it be invoked from cleanup and bg process.
* `store_accounts_unfrozen` which only seems to be invoked from replay and other fg processes.

If this is true then we can pass the thread-pool explicitly to `update_index` and pick the right thread-pool in `store_accounts_frozen` and `store_accounts_unfrozen`. Last commit on the PR is implementing this.